Sources/JavaScriptKit/BasicObjects/JSPromise.swift

```diff
@@ -116,123 +56,67 @@ extension JSPromise where Success: ConvertibleToJSValue, Failure: JSError {
-        let closure = JSOneshotClosure { arguments -> JSValue in
-            guard let result = Success.construct(from: arguments[0]) else {
-                fatalError("\(file):\(line): failed to unwrap success value for `then` callback")
-            }
-            return success(result).jsValue()
-        }
-        return .init(unsafe: jsObject.then!(closure).object!)
+    public static func reject(_ reason: JSValue) -> JSPromise {
+        self.init(unsafelyWrapping: Self.constructor.reject!(reason).object!)
     }
 
-    /** Returns a new promise created from chaining the current `self` promise with the `success`
-    closure invoked on sucessful completion of `self`. The returned promise will have a new type
-    equal to the return type of `success`.
+    /** Schedules the `success` closure to be invoked on sucessful completion of `self`.
     */
-    public func then<ResultSuccess: ConvertibleToJSValue, ResultFailure: ConstructibleFromJSValue>(
-        success: @escaping (Success) -> JSPromise<ResultSuccess, ResultFailure>,
-        file: StaticString = #file,
-        line: Int = #line
-    ) -> JSPromise<ResultSuccess, ResultFailure> {
-        let closure = JSOneshotClosure { arguments -> JSValue in
-            guard let result = Success.construct(from: arguments[0]) else {
-                fatalError("\(file):\(line): failed to unwrap success value for `then` callback")
-            }
-            return success(result).jsValue()
+    @discardableResult
+    public func then(success: @escaping (JSValue) -> JSValue) -> JSPromise {
+        let closure = JSOneshotClosure {
+            return success($0[0])
         }
-        return .init(unsafe: jsObject.then!(closure).object!)
+        return JSPromise(unsafelyWrapping: jsObject.then!(closure).object!)
     }
-}
 
-extension JSPromise where Failure: ConstructibleFromJSValue {
-    /** Returns a new promise created from chaining the current `self` promise with the `failure`
-    closure invoked on rejected completion of `self`. The returned promise will have a new `Success`
-    type equal to the return type of the callback, while the `Failure` type becomes `Never`.
+    /** Schedules the `success` closure to be invoked on sucessful completion of `self`.
     */
-    public func `catch`<ResultSuccess: ConvertibleToJSValue>(
-        failure: @escaping (Failure) -> ResultSuccess,
-        file: StaticString = #file,
-        line: Int = #line
-    ) -> JSPromise<ResultSuccess, Never> {
-        let closure = JSOneshotClosure { arguments -> JSValue in
-            guard let error = Failure.construct(from: arguments[0]) else {
-                fatalError("\(file):\(line): failed to unwrap error value for `catch` callback")
-            }
-            return failure(error).jsValue()
+    @discardableResult
+    public func then(success: @escaping (JSValue) -> JSValue,
+                     failure: @escaping (JSValue) -> JSValue) -> JSPromise {
```

Comment:
```suggestion
    public func then(success: @escaping (JSValue) -> JSValueConvertible,
                     failure: @escaping (JSValue) -> JSValueConvertible) -> JSPromise {
```

I wonder if there’s a way to implement this as `Result<JSValue, JSValue>`? Maybe a disfavored overload so the user has to explicitly specify the label?